lib/evm/instruction/impl.ex

```diff
@@ -50,240 +51,208 @@ defmodule EVM.Instruction.Impl do
+  def div([s0, s1], _) do
+    Integer.floor_div(s0, s1)
+      |> wrap_int
+      |> encode_signed
   end
 
   @doc """
   Signed integer division operation (truncated).
-
-  TODO: Implement opcode
-
-  ## Examples
-
-      iex> EVM.Instruction.Impl.sdiv([], %{stack: []})
-      :unimplemented
   """
   @spec sdiv(stack_args, vm_map) :: op_result
-  def sdiv(_args, %{stack: _stack}) do
-    :unimplemented
+  def sdiv([s0, s1], _) do
+    case decode_signed(s1) do
+      0 ->
+          0
+      1 ->
+        s0
+      -1 ->
+        -decode_signed(s0)
+      _ ->
+        MathHelper.round_int((decode_signed(s0) / decode_signed(s1)))
+    end
   end
 
   @doc """
   Modulo remainder operation.
-
-  TODO: Implement opcode
-
-  ## Examples
-
-      iex> EVM.Instruction.Impl.mod([], %{stack: []})
-      :unimplemented
   """
   @spec mod(stack_args, vm_map) :: op_result
-  def mod(_args, %{stack: _stack}) do
-    :unimplemented
+  def mod([s0, s1], _) do
+    if (s1 == 0), do: 0, else: rem(s0, s1)
   end
 
   @doc """
   Signed modulo remainder operation.
-
-  TODO: Implement opcode
-
-  ## Examples
-
-      iex> EVM.Instruction.Impl.smod([], %{stack: []})
-      :unimplemented
   """
   @spec smod(stack_args, vm_map) :: op_result
-  def smod(_args, %{stack: _stack}) do
-    :unimplemented
+  def smod([_, s1], _) when s1 == 0, do: 0
+  def smod([s0, s1], _) do
+    rem(decode_signed(s0), decode_signed(s1))
   end
 
   @doc """
   Modulo addition operation.
-
-  TODO: Implement opcode
-
-  ## Examples
-
-      iex> EVM.Instruction.Impl.addmod([], %{stack: []})
-      :unimplemented
   """
   @spec addmod(stack_args, vm_map) :: op_result
-  def addmod(_args, %{stack: _stack}) do
-    :unimplemented
-    # stack |> push( ( s0 + s1 ) mod s2 )
+  def addmod([_, _, s2], _) when s2 == 0, do: 0
+  def addmod([s0, s1, s2], _) do
+    rem(s0 + s1, s2)
   end
 
   @doc """
   Modulo multiplication operation.
-
-  TODO: Implement opcode
-
-  ## Examples
-
-      iex> EVM.Instruction.Impl.mulmod([], %{stack: []})
-      :unimplemented
   """
   @spec mulmod(stack_args, vm_map) :: op_result
-  def mulmod(_args, %{stack: _stack}) do
-    :unimplemented
+  def mulmod([_, _, s2], _) when s2 == 0, do: 0
+  def mulmod([s0, s1, s2], _) do
+    rem(s0 * s1, s2)
   end
 
   @doc """
   Exponential operation
 
-  TODO: Implement opcode
-
   ## Examples
 
-      iex> EVM.Instruction.Impl.exp([2, 3], %{stack: []})
-      %{stack: [8]}
-
-      iex> EVM.Instruction.Impl.exp([-2, 7], %{stack: []})
-      %{stack: [-128]}
+      iex> EVM.Instruction.Impl.exp([2, 3], %{})
+      8
 
-      iex> EVM.Instruction.Impl.exp([2, 257], %{stack: []})
-      %{stack: [0]}
+      iex> EVM.Instruction.Impl.exp([2, 257], %{})
+      0
   """
   @spec exp(stack_args, vm_map) :: op_result
-  def exp([s0, s1], %{stack: stack}) do
-    stack |> push(round(:math.pow(s0, s1)))
+  def exp([s0, s1], _) do
+    :crypto.mod_pow(s0, s1, EVM.max_int()) |> :crypto.bytes_to_integer
```

Comment:
Curious how `:crypto.bytes_to_integer` differs from `:binary.decode_unsigned/1`?

Reply:
huh, I don't think it is. I switched it to `:binary.decode_unsigned/1` for consistency